types/i18next/index.d.ts

```diff
@@ -1,157 +1,601 @@
-        parseMissingKeyHandler?: (key: string) => void;
-        appendNamespaceToMissingKey?: boolean;
-        postProcess?: string | any[];
-        returnNull?: boolean;
-        returnEmptyString?: boolean;
-        returnObjects?: boolean;
-        returnedObjectHandler?: (key: string, value: string, options: any) => void;
-        joinArrays?: string;
-        overloadTranslationOptionHandler?: (args: any[]) => TranslationOptions;
-        interpolation?: InterpolationOptions;
-        detection?: any;
-        backend?: any;
-        cache?: any;
+    type Callback = (error: any, t: TranslationFunction) => void;
+
+    type TranslationFunction<TResult = string, TValues extends object = object, TKeys extends string = string> = (key: TKeys | TKeys[], options?: TranslationOptions<TValues>) => TResult;
+
+    interface Resource {
+        [language: string]: ResourceLanguage;
     }
 
-    // init options for react-i18next
-    interface ReactOptions {
-      wait?: boolean;
+    interface ResourceLanguage {
+        [namespace: string]: ResourceKey;
     }
 
-    type TranslationFunction = (key: string, options?: TranslationOptions) => string;
-    type LoadCallback = (error: any, t: TranslationFunction) => void;
-
-    interface I18n {
-        // api
-        init(options?: Options & ReactOptions, callback?: LoadCallback): I18n;
-        use(module: any): I18n;
-        t(key: string, options?: TranslationOptions): string | any | any[];
-        exists(key: string, options?: TranslationOptions): boolean;
-        getFixedT(lng?: string, ns?: string | string[]): TranslationFunction;
-        changeLanguage(lng: string, callback?: LoadCallback): void;
+    interface ResourceKey {
+        [key: string]: any;
+    }
+
+    interface i18n {
+        /**
+         * The default export of the i18next module is an i18next instance ready to be initialized by calling init.
+         * You can create additional instances using the createInstance function.
+         *
+         * @param options - Initial options.
+         * @param callback - will be called after all translations were loaded or with an error when failed (in case of using a backend).
+         */
+        init(options: InitOptions, callback?: Callback): i18n;
+        init(callback?: Callback): i18n;
+
+        loadResources(callback?: (err: any) => void): void;
+
+        /**
+         * The use function is there to load additional plugins to i18next.
+         * For available module see the plugins page and don't forget to read the documentation of the plugin.
+         */
+        use(module: any): i18n;
+
+        /**
+         * Please have a look at the translation functions like interpolation, formatting and plurals for more details on using it.
+         */
+        t<TResult = string, TValues extends object = object, TKeys extends string = string>(key: TKeys | TKeys[], options?: TranslationOptions<TValues>): TResult;
```

Comment:
@GiedriusGrabauskas @DovydasNavickas please notice that `T` does not set result type, it just sets upper type bound.
Look at [TypeScript playground 1](https://www.typescriptlang.org/play/#src=interface%20TranslationFunction%3CT%3E%20%7B%0D%0A%20%20%20%20%3CTResult%20extends%20T%20%3D%20T%2C%20TKey%20extends%20string%20%3D%20string%3E(key%3A%20TKey)%3A%20TResult%0D%0A%7D%0D%0A%0D%0Alet%20t%3A%20TranslationFunction%3Cany%3E%20%3D%20(key)%20%3D%3E%20key%0D%0Alet%20b%3A%20TranslationFunction%3Cboolean%3E%20%3D%20(key)%20%3D%3E%20key%20%3D%3D%3D%20%22test%22%0D%0A%0D%0Aalert(t%3Cstring%2C%20%22key1%22%20%7C%20%22key2%22%3E(%22key1%22).length)%0D%0A%0D%0Aif%20(b%3Cboolean%2C%20string%3E(%22test%22))%20%7B%0D%0A%20%20%20%20alert(%22Success!%22)%0D%0A%7D%20). You can still specify type parameters when calling `t()` function (I simplified example a little).
Also current implementation returns from `getFixedT` `TranslationFunction` with default type arguments - `(key: string | string[], options?: TranslationOptions<object>) => any` and does not allow to specify type arguments when calling this function.